### PR TITLE
[QI2-1176] Update authentication with information retrieved from authconfig endpoint

### PR DESCRIPTION
**Pull request by Mythir**

### CODE REVIEW

**Review by eliasKA — DISMISSED** [review]
Looks good ✅

File: quantuminspire/cli/command_list.py [hunk 1]
```diff
@@ -334,6 +335,10 @@ def get_final_results(job_id: int = typer.Argument(..., help="The id of the run"
 @app.command("login")
 def login(
     host: Optional[str] = typer.Argument(None, help="The URL of the platform to which to connect"),
+    use_local_auth_config: bool = typer.Option(
```
Comment (oschusler) [inline quantuminspire/cli/command_list.py]: I do know that the Zen says "Explicit is better than implicit.", but it also says "Readability counts". I think this is a very length flag. We don't need the `use_` right? or maybe `override_auth_config`?

File: quantuminspire/util/configuration.py [hunk 2]
```diff
@@ -20,15 +20,7 @@
...
-  "auths": {
-    "https://staging.qi2.quantum-inspire.com": {
-      "client_id": "Yz7ni9PUAyT43eUASZfmc1yqI66QxLUJ",
-      "well_known_endpoint":  "https://quantum-inspire-staging.eu.auth0.com/.well-known/openid-configuration"
-    },
-    "https://api.qi2.quantum-inspire.com": {
-      "well_known_endpoint":  "https://auth.qi2.quantum-inspire.com/realms/oidc_production/.well-known/openid-configuration"
-    }
-  }
+  "auths": {}
```
Comment (oschusler) [inline quantuminspire/util/configuration.py]: Should the `DEFAULT_CONFIG` even be here?

File: tests/cli/test_command_list.py [hunk 3]
```diff
@@ -141,11 +141,20 @@ def test_final_results_get_no_results(mocker: MockerFixture) -> None:
...
     store_tokens = mocker.patch("quantuminspire.cli.command_list.Settings.store_tokens")
-    result = runner.invoke(app, ["login", "https://host"])
+    fetch_auth = mocker.patch("quantuminspire.cli.command_list.Settings.fetch_suggested_auth_settings")
+
+    if use_local_auth_config:
+        result = runner.invoke(app, ["login", "https://host", "--use-local-auth-config"])
+        fetch_auth.assert_not_called()
+    else:
+        result = runner.invoke(app, ["login", "https://host"])
+        fetch_auth.assert_called_once()
```
Comment (oschusler) [inline tests/cli/test_command_list.py]: It feels weird to have a parameterized test, where you need an if-statement to test both behaviours. I also don't think we should include the CLI behaviour on the flag in a test whether login succeeds. With or without the flag, you still open the browser. You just do a request to an endpoint, or read a local file.

File: quantuminspire/util/configuration.py [hunk 4]
```diff
@@ -150,6 +142,18 @@ def settings_customise_sources(  # pylint: disable=R0913
             file_secret_settings,
         )
 
+    async def fetch_suggested_auth_settings(self, host: Optional[Url] = None) -> None:
```
Comment (oschusler) [inline quantuminspire/util/configuration.py]: It's not really meant as "suggested". These are the correct settings for this API. However, if you really really have a different opinion...

**Review by oschusler — CHANGES_REQUESTED** [review]

**Comment by oschusler** [comment]
Could you please change the name of the PR? The commit message should reflect what the commit does. It doesn't do "auth config"

File: quantuminspire/cli/command_list.py (hunk 1 above)
Comment (eliasKA) [reply quantuminspire/cli/command_list.py]: I agree I like `override_auth_config` better

File: quantuminspire/util/configuration.py (hunk 2 above)
Comment (eliasKA) [reply quantuminspire/util/configuration.py]: It is being used as skeleton for the auth_config file if this file had not yet been created (see [ensure_config_file_exists](https://github.com/QuTech-Delft/quantuminspire2/blob/4523f758ad5814f81b21c0e2ac3c62b5eced5f30/quantuminspire/util/configuration.py#L28) method). ~~Later on when we want to populate it we expect at least the "auths" key to be present that's why we still need it.~~

Comment (eliasKA) [reply quantuminspire/util/configuration.py]: Ah nevermind I think we can drop it

File: quantuminspire/util/configuration.py (hunk 4 above)
Comment (eliasKA) [reply quantuminspire/util/configuration.py]: Agree, changed to `fetch_auth_settings` instead

File: tests/cli/test_command_list.py (hunk 3 above)
Comment (eliasKA) [reply tests/cli/test_command_list.py]: Yes but to reach full coverage we have to cover both cases:
```
    if not override_auth_config:
        asyncio.run(settings.fetch_auth_settings(host_url))
 ```
 
 Thats why I included both cases here. Reason I decided for one paremterized test is splitting them in two would end up with mostly duplicate code

Comment (eliasKA) [reply tests/cli/test_command_list.py]: Actually I think I get your point now, have a seperate test for testing the `test_override_auth_config` flag, and leave the `login` without including this logic. I will make the change

**Comment by eliasKA** [comment]
@oschusler had another look at `test_configuration.py`
```
    await settings.fetch_auth_settings(host)
    auth_settings = settings.auths[settings.default_host] if host is None else settings.auths[host]
```

This should not be an issue and does not relate to the change of dropping the `DEFAULT_CONFIG` in the file. Whats happening is the following:

We check the behaviour of `fetch_auth_settings` in two cases:
1. When a host is provided (ie a user provides a host when doing a `login`)
2. When a host is not provided (ie a user does **not** provide a host when doing a `login`) in this case it should default to the `default_host: Url = "https://api.qi2.quantum-inspire.com"` hardcoded in the settings

**Review by oschusler — APPROVED** [review]

**Pull request merged**